tests/delete/tests.py

```diff
@@ -800,3 +818,187 @@ def test_fast_delete_full_match(self):
         with self.assertNumQueries(1):
             User.objects.filter(~Q(pk__in=[]) | Q(avatar__desc="foo")).delete()
         self.assertFalse(User.objects.exists())
+
+
+class DatabaseLevelOnDeleteTests(TestCase):
+    def test_deletion_on_nested_cascades(self):
+        foo = Foo.objects.create()
+        bar = Bar.objects.create(foo=foo)
+        baz = Baz.objects.create(bar=bar)
+
+        foo.delete()
+
+        with self.assertRaises(Bar.DoesNotExist):
+            bar.refresh_from_db()
+
+        with self.assertRaises(Baz.DoesNotExist):
+            baz.refresh_from_db()
```

Comment:
Use `count()` or `exists()` instead of catching errors:
```suggestion
        bar = Bar.objects.create(foo=foo)
        baz = Baz.objects.create(bar=bar)
        foo.delete()
        self.assertIs(Bar.objects.filter(pk=bar.pk).exists(), False)
        self.assertIs(Baz.objects.filter(pk=baz.pk).exists(), False)
```